File: app/src/main/java/remix/myplayer/request/network/ApiService.java
package remix.myplayer.request.network;

import androidx.annotation.Nullable;
import io.reactivex.Observable;
import io.reactivex.Single;
import okhttp3.ResponseBody;
import remix.myplayer.BuildConfig;
import remix.myplayer.bean.github.Release;
import retrofit2.http.GET;
import retrofit2.http.Headers;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;

/**
 * Created by Remix on 2017/11/20.
 */

public interface ApiService {

  @POST("search/pc")
  @Headers("User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36")
  Observable<ResponseBody> getNeteaseSearch(@Query("s") String key, @Query("offset") int offset,
      @Query("limit") int limit, @Query("type") int type);

  @GET("song/lyric")
  @Headers("User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36")
  Observable<ResponseBody> getNeteaseLyric(@Query("os") String os, @Query("id") int id,
      @Query("lv") int lv, @Query("kv") int kv, @Query("tv") int tv);

  @GET("search")
  Observable<ResponseBody> getKuGouSearch(@Query("ver") int ver, @Query("man") String man,
      @Query("client") String client,
      @Query("keyword") String keyword, @Query("duration") long duration,
      @Query("hash") String hash);

  @GET("download")
  Observable<ResponseBody> getKuGouLyric(@Query("ver") int ver, @Query("client") String client,
      @Query("fmt") String fmt, @Query("charset") String charSet,
      @Query("id") int id, @Query("accesskey") String accessKey);

  @GET("soso/fcgi-bin/client_search_cp")
  Observable<ResponseBody> getQQSearch(@Query("n") int n,
      @Query("w") String w,
      @Query("format") String format);

  @GET("lyric/fcgi-bin/fcg_query_lyric_new.fcg")
  @Headers("Referer: https://y.qq.com/portal/player.html")
  Observable<ResponseBody> getQQLyric(@Query("songmid") String songmid,
      @Query("g_tk") int g_tk,
      @Query("format") String format,
      @Query("nobase64") int nobase64);

  String BASE_QUERY_PARAMETERS = "?format=json&autocorrect=1&api_key=" + BuildConfig.LASTFM_API_KEY;

  @GET(BASE_QUERY_PARAMETERS + "&method=album.getinfo")
  Observable<ResponseBody> getAlbumInfo(@Query("album") String albumName,
      @Query("artist") String artistName, @Nullable @Query("lang") String language);

  @GET(BASE_QUERY_PARAMETERS + "&method=artist.getinfo")
  Observable<ResponseBody> getArtistInfo(@Query("artist") String artistName,
      @Nullable @Query("lang") String language);

  @GET("repos/{owner}/{repo}/releases/latest")
  @Headers({"token: " + BuildConfig.GITHUB_SECRET_KEY})
  Single<Release> getLatestRelease(@Path("owner") String owner, @Path("repo") String repo);
}

File: app/src/main/java/remix/myplayer/request/network/ApiService.kt
package remix.myplayer.request.network

import io.reactivex.Single
import remix.myplayer.BuildConfig
import remix.myplayer.bean.github.Release
import remix.myplayer.bean.kugou.KLrcResponse
import remix.myplayer.bean.kugou.KSearchResponse
import remix.myplayer.bean.lastfm.LastFmAlbum
import remix.myplayer.bean.lastfm.LastFmArtist
import remix.myplayer.bean.netease.NAlbumSearchResponse
import remix.myplayer.bean.netease.NArtistSearchResponse
import remix.myplayer.bean.netease.NLrcResponse
import remix.myplayer.bean.netease.NSongSearchResponse
import remix.myplayer.bean.netease.NDetailResponse
import remix.myplayer.bean.qq.QLrcResponse
import remix.myplayer.bean.qq.QSearchResponse
import retrofit2.http.*

/**
 * Created by Remix on 2017/11/20.
 */
interface ApiService {
  @GET("repos/{owner}/{repo}/releases/latest")
  @Headers("token: " + BuildConfig.GITHUB_SECRET_KEY)
  fun fetchLatestRelease(@Path("owner") owner: String?, @Path("repo") repo: String?): Single<Release>

  //New Api
  @GET("$BASE_QUERY_PARAMETERS&method=album.getinfo")
  fun searchLastFMAlbum(@Query("album") albumName: String?,
                        @Query("artist") artistName: String?, @Query("lang") language: String?): Single<LastFmAlbum>

  @GET("$BASE_QUERY_PARAMETERS&method=artist.getinfo")
  fun searchLastFMArtist(@Query("artist") artistName: String?,
                         @Query("lang") language: String?): Single<LastFmArtist>

  @GET("search/get")
  @Headers("User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36")
  fun searchNeteaseSong(@Query("s") key: String?, @Query("offset") offset: Int,
                        @Query("limit") limit: Int, @Query("type") type: Int): Single<NSongSearchResponse>

  @GET("search/get")
  @Headers("User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36")
  fun searchNeteaseAlbum(@Query("s") key: String?, @Query("offset") offset: Int,
                         @Query("limit") limit: Int, @Query("type") type: Int): Single<NAlbumSearchResponse>

  @GET("search/get")
  @Headers("User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36")
  fun searchNeteaseArtist(@Query("s") key: String?, @Query("offset") offset: Int,
                          @Query("limit") limit: Int, @Query("type") type: Int): Single<NArtistSearchResponse>

  @GET("song/lyric")
  @Headers("User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36")
  fun searchNeteaseLyric(@Query("os") os: String?, @Query("id") id: Int,
                         @Query("lv") lv: Int, @Query("kv") kv: Int, @Query("tv") tv: Int): Single<NLrcResponse>

  @GET("song/detail")
  @Headers("User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36")
  fun searchNeteaseDetail( @Query("id") id: Int, @Query("ids") ids: String?): Single<NDetailResponse>

  @GET("search")
  fun searchKuGou(@Query("ver") ver: Int, @Query("man") man: String?,
                  @Query("client") client: String?,
                  @Query("keyword") keyword: String?, @Query("duration") duration: Long,
                  @Query("hash") hash: String?): Single<KSearchResponse>

  @GET("download")
  fun searchKuGouLyric(@Query("ver") ver: Int, @Query("client") client: String?,
                       @Query("fmt") fmt: String?, @Query("charset") charSet: String?,
                       @Query("id") id: Int, @Query("accesskey") accessKey: String?): Single<KLrcResponse>

  @GET("soso/fcgi-bin/client_search_cp")
  fun searchQQ(@Query("n") n: Int,
               @Query("w") w: String?,
               @Query("format") format: String?): Single<QSearchResponse>

  @GET("lyric/fcgi-bin/fcg_query_lyric_new.fcg")
  @Headers("Referer: https://y.qq.com/portal/player.html")
  fun searchQQLyric(@Query("songmid") songmid: String?,
                    @Query("g_tk") g_tk: Int,
                    @Query("format") format: String?,
                    @Query("nobase64") nobase64: Int): Single<QLrcResponse>

  companion object {
    const val BASE_QUERY_PARAMETERS = "?format=json&autocorrect=1&api_key=" + BuildConfig.LASTFM_API_KEY
  }
}
